### ISSUE TITLE
preview-tui does not work on MacOS

### ISSUE THREAD

**Issue by tastyminerals** [jarun/nnn#895]
Cannot make `preview-tui` work on MacOS.
Downloaded `preview-tui` and put it into `.config/nnn/pluings`
Installed **tmux** as a dependency via MacPorts and all other additional dependencies.
Updated fish shell config to include
```
set -x NNN_OPENER /Users/pavels/.config/nnn/plugins/nuke
set -x NNN_PLUG 'p:preview-tui'
set -x NNN_FIFO /tmp/nnn.fifo
set -x TERMINAL (which tmux)
```
I followed the `preview-tui` readme and run `nnn -a`.
When in nnn I type `;p` to invoke the plugin but the window just blinks for a millisecond and nothing happens :(

**Comment by luukvbaal** [comment]
Can you test if [this](https://raw.githubusercontent.com/jarun/nnn/4d77066b621e6c0a8fb17e384eba822f5d08248d/plugins/preview-tui) version works?

**Comment by luukvbaal** [comment]
I actually went through the trouble of installing [OSX-KVM](https://github.com/kholia/OSX-KVM) on my laptop and could not reproduce the issue. `preview-tui` from master works fine on that hackintosh VM inside `tmux`. 

Are you sure you're acutally running `tmux`, have `TERMINAL` set to your terminal inside your shell profile, or have `xterm` installed? If any of those three are true the plugin should work.

**Comment by luukvbaal** [comment]
https://github.com/jarun/nnn/issues/617 seems to be a similar issue, see if the discussion in there can't help you figure it out. Closing as it's not an issue with `nnn` but with your local environment.

**Comment by tastyminerals** [comment]
Emm, maybe I am missing something but the only takeaway I could get from the linked issue is to set `TERM=xterm-256color` which did nothing in my case 🤷

**Comment by luukvbaal** [comment]
Did you test the previous version I linked? Like I said I couldn't reproduce the issue on the VM so I assume its something on your end but maybe I'm wrong and there was an issue introduced in later versions of the script.

**Comment by luukvbaal** [comment]
If it doesn't, can you see any errors? Repeatedly pressing `;p` may allow you to read them better if there are any. Please post the output of `env`.

**Comment by tastyminerals** [comment]
Well, the flickering seems to go away after I set `TERM=xterm-256color` so I can't even see any message now.

Here is what I do. Run iTerm2. Type **tmux**, in tmux run `nnn -a`, hit `;p`. Nothing happens.

```
~ tree ~/.config/nnn/plugins/
.config/nnn/plugins/
├── nuke
└── preview-tui
```

`env`

```
COLORFGBG=7;0
COLORTERM=truecolor
COMMAND_MODE=unix2003
HOME=/Users/pavels
ITERM_PROFILE=Default
ITERM_SESSION_ID=w1t1p0:F9844B6D-653F-44C3-908C-CF4CDC1929E1
JAVA_HOME=/Library/Java/JavaVirtualMachines/jdk1.8.0_281.jdk/Contents/Home
LC_CTYPE=UTF-8
LC_TERMINAL=iTerm2
LC_TERMINAL_VERSION=3.4.4
LOGNAME=pavels
NNN_FIFO=/tmp/nnn.fifo
NNN_OPENER=/Users/pavels/.config/nnn/plugins/nuke
NNN_PLUG=p:preview-tui;n:nuke
OMF_CONFIG=/Users/pavels/.config/omf
OMF_PATH=/Users/pavels/.local/share/omf
PATH=/Applications/MacVim.app/Contents/bin:/opt/local/bin:/opt/local/sbin:/usr/local/opt/coreutils/libexec/gnuman:/usr/local/opt/coreutils/libexec/gnubin:/Applications/MacVim.app/Contents/bin:/opt/local/bin:/opt/local/sbin:/Users/pavels/.poetry/bin:/usr/local/bin:/usr/bin:/bin:/usr/sbin:/sbin:/Library/Apple/usr/bin
PWD=/Users/pavels/Dev/Gini/pia_extractor_service
SHELL=/opt/local/bin/fish
SHLVL=1
SSH_AUTH_SOCK=/private/tmp/com.apple.launchd.7z2eCISFyy/Listeners
TERM=xterm-256color
TERMINAL=/opt/local/bin/tmux
TERM_PROGRAM=iTerm.app
TERM_PROGRAM_VERSION=3.4.4
TERM_SESSION_ID=w1t1p0:F9844B6D-653F-44C3-908C-CF4CDC1929E1
TMPDIR=/var/folders/j2/q84fjhdd4_938763t_6yk8xw0000gn/T/
USER=pavels
XPC_FLAGS=0x0
XPC_SERVICE_NAME=0
__CFBundleIdentifier=com.googlecode.iterm2
__CF_USER_TEXT_ENCODING=0x1F5:0x0:0x2
```

`cat ~/.config/fish/fish.config`

```
# Set environment vars for Fish shell instance.
set -xg PATH /opt/local/bin /opt/local/sbin $PATH
# Allow mvim to be run from terminal
set -xg PATH /Applications/MacVim.app/Contents/bin $PATH
set -x JAVA_HOME (/usr/libexec/java_home -v 1.8.0_281)
set -x NNN_OPENER /Users/pavels/.config/nnn/plugins/nuke
set -x NNN_PLUG 'p:preview-tui;n:nuke'
set -x NNN_FIFO /tmp/nnn.fifo
set -x TERMINAL (which tmux)
set -x TERM xterm-256color

# Set aliases
alias python="/usr/bin/python3.8"
alias python2="/usr/bin/python"

alias rm="rm -i"
alias cp="cp -i"

# Fish shell misc configs.
test -e {$HOME}/.iterm2_shell_integration.fish ; and source {$HOME}/.iterm2_shell_integration.fish

# Enable jump for shell
status --is-interactive; and source (jump shell fish | psub)

function __jump_add --on-variable PWD
  status --is-command-substitution; and return
  jump chdir
end

function __jump_hint
  set -l term (string replace -r '^j ' '' -- (commandline -cp))
  jump hint "$term"
end

function j
  set -l dir (jump cd "$argv")
  test -d "$dir"; and cd "$dir"
end

# config for jump
complete --command j --exclusive --arguments '(__jump_hint)'

fish_vi_key_bindings
```

**Comment by luukvbaal** [comment]
Output of `tmux -V`?

**Comment by luukvbaal** [comment]
Anyways I had never used OSX before but `nnn` with `preview-tui` worked out of the box after installing `brew` and running `brew install tmux; brew install nnn; brew install tree`
Again this is on a VM so unless anyone else actually running OSX can confirm the issue I'd suggest you try to figure out your environment, good luck.

**Comment by tastyminerals** [comment]
> Output of `tmux -V`?

tmux 3.1c

**Comment by luukvbaal** [comment]
I'm sorry I can't tell why it wouldn't work for you, with the information you provided I would say it should work. However like I said I'm unfamiliar with OSX so I'll reopen to see if anyone else can step in...

**Comment by jarun** [comment]
@leovilok any ideas?

**Comment by 0xACE** [comment]
```
env | grep -i tmux
```

You must be within a tmux environment for `preview-tui` to work

**Comment by toddyamakawa** [comment]
@tastyminerals I think there are two problems.
1. It looks like you're not running `nnn` from within `tmux`. The output of `env` doesn't show the `TMUX` variable being set. Before you run `nnn`, make sure you run `tmux`. When you're inside of tmux, run `echo $TMUX` to make sure it prints something. Then you can run `nnn` and `preview-tui`
2. The second problem is that your `TERMINAL` variable is actually set to the path of tmux. This means that when you run `preview-tui`, it can't open tmux or kitty so it tries to open whatever your `$TERMINAL` variable is set to. Since you have it set to your tmux path, it tries to run tmux and fails because tmux isn't a terminal.

**Comment by luukvbaal** [comment]
> The output of `env` doesn't show the `TMUX` variable being set. 
> The second problem is that your `TERMINAL` variable is actually set to the path of tmux. 

I noticed both of these as well but he said he was running tmux:
> Here is what I do. Run iTerm2. Type tmux, in tmux run nnn -a, hit ;p. Nothing happens.
 
and `TERMINAL` containing the `tmux` path would be overwritten by
`if [ -e "${TMUX%%,*}" ] && tmux -V | grep -q '[ -][3456789]\.'; then
TERMINAL=tmux` if he's actually running `preview-tui` in `tmux`.

**Comment by toddyamakawa** [comment]
> if he's actually running `preview-tui` in `tmux`.

The behavior they're seeing is consistent with `nnn` not being run in `tmux` though. I can't think of another case where the "the window just blinks and nothing happens"

@tastyminerals quick clarification. When you say "the window just blinks" are you saying that you see the tmux split being created and then immediately closed?

**Comment by leovilok** [comment]
@tastyminerals as 0xACE and toddyamakawa said, you're not actually running tmux (installing tmux is not the same as running it). Putting `set -x TERMINAL (which tmux)` into your env, you're tricking the plugin into thinking that you're in a tmux session, while you're not actually in one.

See e.g. https://github.com/fish-shell/fish-shell/issues/4434 on how to start tmux automatically when you start a shell session.

Basically you can either add something like this to your `config.fish`:
```fish
if status is-interactive
and not set -q TMUX
    exec tmux
end
```

Or setup iterm to start `tmux` instead of `fish`.

**Comment by tastyminerals** [comment]
I changed `TERMINAL` variable to `xterm`. Here is what I do:

![ezgif-2-a3bac4a18121](https://user-images.githubusercontent.com/7676160/111882814-e6416d00-89b7-11eb-9f64-0ee190030d4d.gif)

**Comment by tastyminerals** [comment]
> > if he's actually running `preview-tui` in `tmux`.
> 
> The behavior they're seeing is consistent with `nnn` not being run in `tmux` though. I can't think of another case where the "the window just blinks and nothing happens"
> 
> @tastyminerals quick clarification. When you say "the window just blinks" are you saying that you see the tmux split being created and then immediately closed?

No, I didn't notice any splits. The blinking doesn't happen if I run nnn from tmux but it does happen 1/3 times if I run nnn from iTerm2.

**Comment by leovilok** [comment]
Can you show the env that you have inside nnn ? (press <key>!</key> in nnn, then call `env` in the shell).

**Comment by jarun** [comment]
@tastyminerals also, do you have the plugins installed in `${XDG_CONFIG_HOME:-$HOME/.config}/nnn/plugins`?

**Comment by ranelpadon** [comment]
Same issue on my end, I'm using macOS Mojave + `Alacritty` + `zsh` + `tmux 3.1c`, it flickers initially, then will have this error:
<img width="1440" alt="Screen Shot 2021-03-21 at 8 15 23 AM" src="https://user-images.githubusercontent.com/4292088/111889146-dd838200-8a1d-11eb-8323-c08f025bd3a1.png">

Plugins are installed:
<img width="484" alt="Screen Shot 2021-03-21 at 8 19 14 AM" src="https://user-images.githubusercontent.com/4292088/111889207-6ac6d680-8a1e-11eb-9379-c8babdce6983.png">



and I have these settings:
<img width="497" alt="Screen Shot 2021-03-21 at 8 20 13 AM" src="https://user-images.githubusercontent.com/4292088/111889193-508cf880-8a1e-11eb-8d7a-b335ad4cd07a.png">

**Comment by ranelpadon** [comment]
Here's my `env` inside `nnn`:
```sh
ALACRITTY_LOG=/var/folders/57/9zlb1yfd703891dbnmlgcbtm0000gn/T/Alacritty-89921.log
Apple_PubSub_Socket_Render=/private/tmp/com.apple.launchd.ooE7idzow5/Render
BAT_THEME=TwoDark
COLORTERM=truecolor
COVERAGE_SKIP=1
DRUSH_PHP=/Applications/XAMPP/xamppfiles/bin/php
FZF_DEFAULT_COMMAND=     fd     --hidden     --follow     --no-ignore     --exclude node_modules     --exclude logs     --exclude media     --exclude static     --exclude cache     --exclude .git     --exclude '.git'     --exclude '*.pyc'     --exclude '*.compiled'     --exclude '*.scssc'
HOME=/Users/ranelpadon
LANG=en_US.UTF-8
LC_ALL=en_US.UTF-8
LC_CTYPE=en_US.UTF-8
LESS=-R
LOGNAME=ranelpadon
LSCOLORS=Gxfxcxdxbxegedabagacad
OLDPWD=/Users/ranelpadon/Desktop/keybr
P9K_SSH=0
P9K_TTY=old
PAGER=less
PATH=/Users/ranelpadon/google-cloud-sdk/bin:/usr/local/opt/node@12/bin:/usr/local/Cellar/pyenv-virtualenv/1.1.5/shims:/Users/ranelpadon/.pyenv/shims:/Users/ranelpadon/.pyenv/bin:/Users/ranelpadon/.emacs.d/bin:/usr/local/bin:/usr/bin:/bin:/usr/sbin:/sbin:/usr/local/MacGPG2/bin:/Applications/Postgres.app/Contents/Versions/latest/bin:/Users/ranelpadon/google-cloud-sdk/bin:/usr/local/opt/node@12/bin:/usr/local/Cellar/pyenv-virtualenv/1.1.5/shims:/Users/ranelpadon/.pyenv/shims:/Users/ranelpadon/.pyenv/bin:/Users/ranelpadon/.emacs.d/bin:/usr/local/opt/fzf/bin:/Users/ranelpadon/.rvm/bin
PWD=/Users/ranelpadon/Desktop/keybr
PYENV_ROOT=/Users/ranelpadon/.pyenv
PYENV_SHELL=zsh
PYENV_VIRTUALENV_INIT=1
PYTHONIOENCODING=UTF-8
PYTHONSTARTUP=/Users/ranelpadon/.python_startup.py
SHELL=/usr/local/bin/zsh
SHLVL=5
SSH_AUTH_SOCK=/private/tmp/com.apple.launchd.ruwMkIPytJ/Listeners
TERM=screen-256color
TMPDIR=/var/folders/57/9zlb1yfd703891dbnmlgcbtm0000gn/T/
TMUX=/private/tmp/tmux-501/default,6863,0
TMUX_PANE=%6
UPDATE_ZSH_DAYS=60
USER=ranelpadon
XPC_FLAGS=0x0
XPC_SERVICE_NAME=0
ZSH=/Users/ranelpadon/.oh-my-zsh
__CF_USER_TEXT_ENCODING=0x1F5:0x0:0x0
rvm_bin_path=/Users/ranelpadon/.rvm/bin
rvm_path=/Users/ranelpadon/.rvm
rvm_prefix=/Users/ranelpadon
rvm_version=1.27.0 (latest)
NNN_PLUG=p:preview-tui;t:preview-tabbed;d:dragdrop;v:imgview
NNN_BMS=r:"/Users/ranelpadon/Data/UP DGE/Resources"
NNN_FIFO=/tmp/nnn.fifo
_system_type=Darwin
_system_name=OSX
_system_version=10.14
_system_arch=x86_64
NNN_PIPE=/tmp/nnn-pipe.11473
NNNLVL=2
nnn=Screen Shot 2020-12-06 at 5.59.31 AM.png
_=/usr/bin/env
```

**Comment by luukvbaal** [comment]
@tastyminerals is the executable set on `preview-tui` if not try `chmod +x preview-tui`.

**Comment by ranelpadon** [comment]
Sample run:
https://user-images.githubusercontent.com/4292088/111889992-6bf90300-8a20-11eb-812a-969b0c2f2333.mov

you could notice that there's some line numbers being shown triggering the error if you pause the video at some point:
<img width="652" alt="Screen Shot 2021-03-21 at 8 37 36 AM" src="https://user-images.githubusercontent.com/4292088/111890024-bc706080-8a20-11eb-9e0b-9cec3213bbe9.png">

**Comment by ranelpadon** [comment]
And the binaries are executable on my end:
<img width="421" alt="Screen Shot 2021-03-21 at 8 39 23 AM" src="https://user-images.githubusercontent.com/4292088/111890049-03f6ec80-8a21-11eb-8000-ded50ebd4167.png">

**Comment by luukvbaal** [comment]
@ranelpadon you problem is different than @tastyminerals, which version of the script are you using, master?

**Comment by ranelpadon** [comment]
I think my issue is similar/related to @tastyminerals, just installed nnn from brew few hours ago (`3.6`).

**Comment by luukvbaal** [comment]
Please try `./getplugs master`.

**Comment by luukvbaal** [comment]
@ranelpadon I can actually reproduce the `cannot open '100' (no such file or directory)` will try to investigate.
The `integer expression expected` should be fixed on `master` however.

**Comment by ranelpadon** [comment]
Tried `./getplugs master`. Same error though.

**Comment by luukvbaal** [comment]
@ranelpadon Should be fixed in https://github.com/jarun/nnn/pull/900/commits/9a06f846539afe9aa31c6c7c2dd5763693e8e61f.

**Comment by ranelpadon** [comment]
@luukvbaal Thanks, I tried your patch in my local, the error is gone indeed. This is the output now, do I need to install other libraries to preview the image? Hence, PNG/JPG preview will just show the metadata instead of rendering them in the terminal, is that expected or am I missing something? Of course, if I click <kbd>Enter</kbd>, it will use the Mac's native `Preview.app` which is not really I want:
<img width="1440" alt="Screen Shot 2021-03-21 at 9 44 17 AM" src="https://user-images.githubusercontent.com/4292088/111891048-686a7980-8a2a-11eb-9462-e866d2518008.png">

**Comment by luukvbaal** [comment]
The script uses either `ueberzug`, `kitty`, `viu` or `catimg` , please read the optional dependencies at the top of `preview-tui`. `kitty` and `ueberzug` can display the actual file, `viu` and `catimg` are terminal image viewers I'm not sure which of those four, if any, work on OSX.

**Comment by luukvbaal** [comment]
Closing since @ranelpadon confirmed it works on OSX. @tastyminerals I suspect your plugins are not executable...

**Comment by tastyminerals** [comment]
> @tastyminerals is the executable set on `preview-tui` if not try `chmod +x preview-tui`.

This! Now it works. 
The readme does not say anything amount **chmod**ing the plugin file, does it?

Also, on MacOS, xterm needs `DISPLAY` to be set otherwise it throws `xterm: Xt error: Cannot find display ...`

**Comment by jarun** [comment]
Second sentence here: https://github.com/jarun/nnn/tree/master/plugins#introduction

Anyway, if you [install using `getplugs`](https://github.com/jarun/nnn/tree/master/plugins#installation) you wouldn't have to do anything manually.

**Comment by luukvbaal** [comment]
> Also, on MacOS, xterm needs `DISPLAY` to be set otherwise it throws `xterm: Xt error: Cannot find display ...`

Note that `xterm` is not actually a requirement for `preview-tui`, you can keep on using your preferred terminal  with `tmux`.

**Comment by ranelpadon** [comment]
By the way, tried the various image viewers, these are my observations, for others that might have same curiosity:
- `viu` for Mac need to be compiled via Rust build, not easy to install, but doable:
    <img width="1256" alt="Screen Shot 2021-03-21 at 8 33 24 PM" src="https://user-images.githubusercontent.com/4292088/111905740-23753000-8a88-11eb-8eb5-7750c4e047de.png">

    It looks good in `iTerm2`/`Kitty`, but it's pixelated in `Alacritty` for some reason (I think they don't have support for hi-res images yet, or haven't implemented the Graphics Protocol unlike other terminals). I use Alacritty mainly though:
    <img width="1275" alt="Screen Shot 2021-03-21 at 9 00 44 PM" src="https://user-images.githubusercontent.com/4292088/111905824-8f579880-8a88-11eb-80ed-c2b15a923921.png">

    `viu` preview doesn't work though inside of `nnn` preview in all terminals (iTerm, Kitty, and Alacritty):
    <img width="1440" alt="Screen Shot 2021-03-21 at 8 42 43 PM" src="https://user-images.githubusercontent.com/4292088/111905908-068d2c80-8a89-11eb-904b-a4abd622cf0f.png">

- `catimg` is easy to install via `brew`, but it's pixelated to all terminals, even outside of  `tmux`/`nnn`: iTerm, Kitty, and Alacritty. However, `catimg` works inside of `nnn` preview in all terminals (iTerm, Kitty, and Alacritty):
    <img width="1275" alt="Screen Shot 2021-03-21 at 9 00 44 PM" src="https://user-images.githubusercontent.com/4292088/111905927-1e64b080-8a89-11eb-998c-534dd1c65107.png">

Couldn't make `ueberzug` to work, even outside of `tmux`/`nnn`. Other possible image prewiewers to explore in Mac:
https://askubuntu.com/questions/97542/how-do-i-make-my-terminal-display-graphical-pictures#answer-698064

**Comment by jarun** [comment]
> Couldn't make ueberzug to work

What problem did you face?

**Comment by ranelpadon** [comment]
@jarun Do I need to update my code as per https://github.com/jarun/nnn/commit/ac29a7c166591c956cc4224272df429006964542?
<img width="1440" alt="Screen Shot 2021-03-21 at 9 48 43 PM" src="https://user-images.githubusercontent.com/4292088/111907230-5b33a600-8a8f-11eb-9310-6efbf34c458f.png">

**Comment by ranelpadon** [comment]
Did ` ./getplugs master` now, same issue with `ueberzug`.

**Comment by jarun** [comment]
See if googling `fifo "interrupted system call"` helps. I don't use macOS so can't help with local this local issue.

**Comment by luukvbaal** [comment]
> > Couldn't make ueberzug to work
> 
> What problem did you face?

`ueberzug` relies on X11 calls to display the image. As far as I as I can tell `XQuartz` can serve as an interface betweens X11 calls and the MacOs display server. You might try your luck with that. 

I'm curious as well so I'll look into further when I get the chance.

**Comment by jarun** [comment]
If someone gets a taste of `ueberzug` it's very difficult to go back to the other viewing methods.
And I can see the performance is great on `xfce-terminal`+`tmux` with `pillow-simd` and jpeg conversion.

I am also checking with the `dvtm` guys if there's a way to use it. `dvtm` is a far more lightweight and cruft-free terminal multiplexer.

**Comment by luukvbaal** [comment]
@ranelpadon I tried running `ueberzug` on my OSX VM by installing:
```
brew install python
brew install xquartz
brew install libxext
sudo pip3 install ueberzug
```

Then running `ueberzug results in error:

```
Xlib.error.DisplayNameError: Bad display name "/private/tmp/com.apple.launchd.JwjGcIcjbD/org.macosforge.xquartz:0"
```

trying after `export DISPLAY=:0` results in 

```
Xlib.error.DisplayConnectionError: Can't connect to display ":0" [Errno 61] Connection refused`
```

I would say that `python-xlib` needs to support `XQuartz` for `ueberzug` to work on OSX.

**Comment by ranelpadon** [comment]
Thanks for your feedback, putting here some relevant discussion also regarding Ueberzug, Alacritty, and macOS, as future reference also to myself:
https://github.com/ranger/ranger/issues/1839
